src/fp32.rs

```diff
@@ -0,0 +1,296 @@
+// SPDX-License-Identifier: MPL-2.0
+
+//! Finite field arithmetic for any field GF(p) for which p < 2^32.
+
+use crate::fp::{hi32, lo32, MAX_ROOTS};
+
+/// This structure represents the parameters of a finite field GF(p) for which p < 2^32.
+///
+/// See also [`FieldParameters`](crate::fp::FieldParameters).
+#[derive(Debug, PartialEq, Eq)]
+pub(crate) struct FieldParameters32 {
+    /// The prime modulus `p`.
+    pub p: u32,
+    /// `base = 2^32`.
+    pub base: u128,
+    /// `mu = -p^(-1) mod base`.
+    pub mu: u32,
+    /// `r2 = (2^32)^2 mod p`.
+    pub r2: u32,
+    /// The `2^num_roots`-th -principal root of unity. This element is used to generate the
+    /// elements of `roots`.
+    pub g: u32,
+    /// The number of principal roots of unity in `roots`.
+    pub num_roots: usize,
+    /// Equal to `2^b - 1`, where `b` is the length of `p` in bits.
+    pub bit_mask: u32,
+    /// `roots[l]` is the `2^l`-th principal root of unity, i.e., `roots[l]` has order `2^l` in the
+    /// multiplicative group. `roots[0]` is equal to one by definition.
+    pub roots: [u32; MAX_ROOTS + 1],
+}
```

Comment:
Instead of duplicating `FieldParameters`, consider making it generic.

```rust
pub(crate) struct FieldParameters<W> {
    pub p: W,
    pub base: u128,
    pub mu: W,
    pub r2: W,
    pub g: W,
    pub num_roots: usize,
    pub bit_mask: W,
    pub roots: [W; MAX_ROOTS + 1],
}
```

Reply:
This is a good idea, but I propose to work on that version in a different PR.

Reply:
Why not now? I don't think we're in a hurry to land this.

Reply:
See for the implementation that unifies both implementations https://github.com/divviup/libprio-rs/pull/1099